### PR TITLE
Add support for Minitest::SharedExamples

### PR DESCRIPTION
**Pull request by dingn1**
This is adding support for `Minitest::SharedExamples`. In Minitest we can use `SharedExamples` to create extensions inside application, which would make the searching of the `test_method_name ` and `test_path` bad. This is  adding a new pattern of searching when the old one does not work.

### CODE REVIEW

**Comment by ArturT** [comment]
Could you add shared example test in `test_examples/fast/` directory.

In order to use shared example do we have to add another gem as dependency or it's just build in minitest gem?

Could you also update test in spec/knapsack/adapters/minitest_adapter_spec.rb to cover both context for regular test and shared example.

**Comment by dingn1** [comment]
It's just build in minitest gem. I will try update the specs.

**Comment by dingn1** [comment]
@ArturT I'm not so sure about the spec part could you give some guide? I just finished the base line of it. It's like a fake_test that include the `SharedExampleSpec` and should correctly return the test_path for that.

File: test_examples/fast/SharedExampes_test.rb [hunk 1]
```diff
@@ -0,0 +1,28 @@
+require 'test_helper'
+require 'minitest/spec'
+
+class Minitest::SharedExamples < Module
+  include Minitest::Spec::DSL
+end
+
+SharedExampleSpec = Minitest::SharedExamples.new do
```
Comment (ArturT) [inline test_examples/fast/SharedExampes_test.rb]: You defined shared example here. Could you add at the bottom a new describe and include there shared example so we can see if they can be executed and recorded by knapsack (Travis will do that https://github.com/ArturT/knapsack/blob/master/.travis.yml#L62 ).

File: test_examples/fast/SharedExampes_test.rb [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+  def test_no_way
+    sleep 0.2
+    refute_match(/^no/i, 'yes')
+  end
+
+  def test_that_will_be_skipped
+    sleep 1
+    skip 'test this later'
+  end
+
```
Comment (ArturT) [inline test_examples/fast/SharedExampes_test.rb]: don't need this new line

File: spec/knapsack/adapters/minitest_adapter_spec.rb [hunk 3]
```diff
@@ -124,4 +124,27 @@ def name
 
     it { should eq './spec/knapsack/adapters/minitest_adapter_spec.rb' }
   end
+
+  describe '.test_path for shared examles' do
```
Comment (ArturT) [inline spec/knapsack/adapters/minitest_adapter_spec.rb]: typo in examles

**Comment by ArturT** [comment]
You could keep single describe for `.test_path` method. Here is example. 

```ruby
  describe '.test_path' do
    subject { described_class.test_path(obj) }

    before do
      parent_of_test_dir = File.expand_path('../../../', File.dirname(__FILE__))
      parent_of_test_dir_regexp = Regexp.new("^#{parent_of_test_dir}")
      described_class.class_variable_set(:@@parent_of_test_dir, parent_of_test_dir_regexp)
    end

    context 'when regular tes' do
      class FakeUserTest
        def test_user_age; end

        # method provided by Minitest
        # it returns test method name
        def name
          :test_user_age
        end
      end

      let(:obj) { FakeUserTest.new }

      it { should eq './spec/knapsack/adapters/minitest_adapter_spec.rb' }
    end

    context 'when shared example test' do
      class FakeSharedExampleUserTest
        # TODO
      end

      let(:obj) { FakeSharedExampleUserTest.new }

      it { should eq './spec/knapsack/adapters/minitest_adapter_spec.rb' }
    end
  end
```

You need to prepare proper FakeSharedExampleUserTest for your case. You could use binding.pry to bind in .test_path method and see what's the structure of the obj in case of shared example (the shared example you added in test_examples directory).

You can run tests from that directory with
KNAPSACK_TEST_FILE_PATTERN="test_examples/**{,/*/**}/*_test.rb" bundle exec rake "knapsack:minitest[--verbose]"

File: test_examples/fast/SharedExampes_test.rb (hunk 1 above)
Comment (dingn1) [inline test_examples/fast/SharedExampes_test.rb]: Checked locally executed fine.

File: test_examples/fast/SharedExamples_test.rb [hunk 4]
```diff
@@ -0,0 +1,31 @@
+require 'test_helper'
```
Comment (ArturT) [inline test_examples/fast/SharedExamples_test.rb]: please rename this file to shared_examples_test.rb

**Comment by ArturT** [comment]
I noticed you removed your test for `.test_path` method. I can write test for that then.

**Comment by ArturT** [comment]
I added missing spec for minitest adapter https://github.com/ArturT/knapsack/blob/master/spec/knapsack/adapters/minitest_adapter_spec.rb#L129

I've just released knapsack v1.12.0. I also updated pro version gem https://github.com/KnapsackPro/knapsack_pro-ruby ( https://knapsackpro.com )

@dingn1 Thank you for the PR and great work :)

**Comment by dingn1** [comment]
@ArturT Thanks for your help.

**Pull request merged**